frame/staking/src/voter_bags.rs

```diff
@@ -0,0 +1,883 @@
+			// in the event that they included it, we can just pass the iterator through unchanged.
+			Box::new(iter.rev())
+		} else {
+			// otherwise, insert it here.
+			Box::new(iter.chain(iter::once(VoteWeight::MAX)).rev())
+		};
+		iter
+			.filter_map(Bag::get)
+			.flat_map(|bag| bag.iter())
+	}
+
+	/// Insert a new voter into the appropriate bag in the voter list.
+	///
+	/// If the voter is already present in the list, their type will be updated.
+	/// That case is cheaper than inserting a new voter.
+	pub fn insert_as(account_id: &AccountIdOf<T>, voter_type: VoterType) {
+		// if this is an update operation we can complete this easily and cheaply
+		if !Node::<T>::update_voter_type_for(account_id, voter_type) {
+			// otherwise, we need to insert from scratch
+			let weight_of = Pallet::<T>::weight_of_fn();
+			let voter = Voter { id: account_id.clone(), voter_type };
+			Self::insert(voter, weight_of);
+		}
+	}
+
+	/// Insert a new voter into the appropriate bag in the voter list.
+	pub fn insert(voter: VoterOf<T>, weight_of: impl Fn(&T::AccountId) -> VoteWeight) {
+		Self::insert_many(sp_std::iter::once(voter), weight_of);
+	}
+
+	/// Insert several voters into the appropriate bags in the voter list.
+	///
+	/// This is more efficient than repeated calls to `Self::insert`.
+	pub fn insert_many(
+		voters: impl IntoIterator<Item = VoterOf<T>>,
+		weight_of: impl Fn(&T::AccountId) -> VoteWeight,
+	) -> u32 {
+		let mut bags = BTreeMap::new();
+		let mut count = 0;
+
+		for voter in voters.into_iter() {
+			let weight = weight_of(&voter.id);
+			let bag = notional_bag_for::<T>(weight);
+			bags.entry(bag).or_insert_with(|| Bag::<T>::get_or_make(bag)).insert(voter);
+			count += 1;
+		}
+
+		for (_, bag) in bags {
+			bag.put();
+		}
+
+		crate::VoterCount::<T>::mutate(|prev_count| *prev_count = prev_count.saturating_add(count));
+		count
+	}
+
+	/// Remove a voter (by id) from the voter list.
+	pub fn remove(voter: &AccountIdOf<T>) {
+		Self::remove_many(sp_std::iter::once(voter))
+	}
+
+	/// Remove many voters (by id) from the voter list.
+	///
+	/// This is more efficient than repeated calls to `Self::remove`.
+	pub fn remove_many<'a>(voters: impl IntoIterator<Item = &'a AccountIdOf<T>>) {
+		let mut bags = BTreeMap::new();
+		let mut count = 0;
+
+		for voter_id in voters.into_iter() {
+			let node = match Node::<T>::from_id(voter_id) {
+				Some(node) => node,
+				None => continue,
+			};
+			count += 1;
+
+			// clear the bag head/tail pointers as necessary
+			let bag = bags.entry(node.bag_upper).or_insert_with(|| Bag::<T>::get_or_make(node.bag_upper));
+			bag.remove_node(&node);
+
+			// now get rid of the node itself
+			crate::VoterNodes::<T>::remove(voter_id);
+			crate::VoterBagFor::<T>::remove(voter_id);
+		}
+
+		for (_, bag) in bags {
+			bag.put();
+		}
+
+		crate::VoterCount::<T>::mutate(|prev_count| *prev_count = prev_count.saturating_sub(count));
```

Comment:
is it possible this count is not accurate because the remove call doesnt actually remove something?

Reply:
Above, when we fetch `node`, if no node exists then we `continue` before editing `count`. `count` is accurate with respect to nodes, and we have debug assertions sprinkled in various places which help retain the property that nodes stay accurate with respect to nominators and validators.

Reply:
If we change `remove_node` to return true if the node is removed (it was pointing to prev/next or head & tail where pointed at it.) and track the return values from `VoterNodes::<T>::remove(voter_id)` & `VoterBagFor::<T>::remove(voter_id)` we can than at least do a debug assert that all the afermentioned functions return `true`/`Some`